rust/src/services/messages/mod.rs

```diff
@@ -0,0 +1,346 @@
+mod message_parser;
+mod pre_processor;
+mod state_machine;
+
+pub use self::{
+    message_parser::{
+        MessageParserError,
+        MessageParserRequest,
+        MessageParserResponse,
+        MessageParserService,
+    },
+    pre_processor::{
+        PreProcessorError,
+        PreProcessorRequest,
+        PreProcessorResponse,
+        PreProcessorService,
+    },
+    state_machine::{
+        StateMachineError,
+        StateMachineRequest,
+        StateMachineResponse,
+        StateMachineService,
+    },
+};
+
+use std::{
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use futures::{future::poll_fn, Future};
+use thiserror::Error;
+use tower::{Service, ServiceBuilder};
+use tracing_futures::Instrument;
+use tracing_tower::request_span;
+use uuid::Uuid;
+
+use crate::{
+    message::MessageOwned,
+    utils::trace::{Traceable, Traced},
+};
+
+/// Associate an ID to the given request, and attach a span to the request.
+pub fn make_traceable_request<R>(req: R) -> Traced<R> {
+    let id = Uuid::new_v4();
+    let span = error_span!("request", id = ?id);
+    Traced::new(req, span)
+}
+
+/// Return the [`tracing::Span`] associated to the given request.
+fn req_span<R>(req: &Traced<R>) -> tracing::Span {
+    req.span().clone()
+}
+
+/// Decorate the given service with a tracing middleware.
+fn with_tracing<S, R>(service: S) -> TracingService<S, R>
+where
+    S: Service<Traced<R>>,
+{
+    ServiceBuilder::new()
+        .layer(request_span::layer(req_span as for<'r> fn(&'r _) -> _))
+        .service(service)
+}
+
+type TracingService<S, R> = request_span::Service<S, Traced<R>, fn(&Traced<R>) -> tracing::Span>;
```

Comment:
I don't get what that means to be honest😅
Is it an alias for a function?

Reply:
It's a normal type alias. The not so usual part of this type is `fn(&Traced<R>) -> tracing::Span`. It's the type of a function that takes a `&Traced<R>` and returns a `Span`. This is not to be confused with the `Fn`, `FnOnce` and `FnMut` _traits_ (not _types_!), which are implemented by functions.

Reply:
See: https://doc.rust-lang.org/reference/types/function-item.html

Reply:
it just looks a bit unusual here, but what rust calls a *function item type* is equivalent to a simple *function pointer* in other languages.

Reply:
Yeah rust also has function pointers. The difference is somewhat subtle: https://stackoverflow.com/a/54629954/1836144